sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaCommitOffset.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A {@link PTransform} that commits offsets of {@link KafkaRecord}. */
+public class KafkaCommitOffset<K, V>
+    extends PTransform<
+        PCollection<KV<KafkaSourceDescriptor, KafkaRecord<K, V>>>, PCollection<Void>> {
+  private final KafkaIO.ReadSourceDescriptors<K, V> readSourceDescriptors;
+
+  KafkaCommitOffset(KafkaIO.ReadSourceDescriptors<K, V> readSourceDescriptors) {
+    this.readSourceDescriptors = readSourceDescriptors;
+  }
+
+  static class CommitOffsetDoFn extends DoFn<KV<KafkaSourceDescriptor, Long>, Void> {
+    private static final Logger LOG = LoggerFactory.getLogger(CommitOffsetDoFn.class);
+    private final Map<String, Object> offsetConsumerConfig;
+    private final Map<String, Object> consumerConfig;
+    private final SerializableFunction<Map<String, Object>, Consumer<byte[], byte[]>>
+        consumerFactoryFn;
+
+    private transient ConsumerSpEL consumerSpEL = null;
+
+    CommitOffsetDoFn(KafkaIO.ReadSourceDescriptors readSourceDescriptors) {
+      offsetConsumerConfig = readSourceDescriptors.getOffsetConsumerConfig();
+      consumerConfig = readSourceDescriptors.getConsumerConfig();
+      consumerFactoryFn = readSourceDescriptors.getConsumerFactoryFn();
+    }
+
+    @ProcessElement
+    public void processElement(@Element KV<KafkaSourceDescriptor, Long> element) {
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, element.getKey());
+      try (Consumer<byte[], byte[]> offsetConsumer =
+          consumerFactoryFn.apply(
+              KafkaIOUtils.getOffsetConsumerConfig(
+                  "commitOffset", offsetConsumerConfig, updatedConsumerConfig))) {
+        try {
+          offsetConsumer.commitSync(
+              Collections.singletonMap(
+                  element.getKey().getTopicPartition(),
+                  new OffsetAndMetadata(element.getValue() + 1)));
+        } catch (Exception e) {
+          // TODO: consider retrying.
+          LOG.warn("Getting exception when committing offset: {}", e.getMessage());
+        }
+      }
+    }
+
+    private Map<String, Object> overrideBootstrapServersConfig(
+        Map<String, Object> currentConfig, KafkaSourceDescriptor description) {
+      checkState(
+          currentConfig.containsKey(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG)
+              || description.getBootStrapServers() != null);
```

Comment:
Yes, that's possible since we no longer force the user to provide bootstrap servers when constructing transforms. The bootstrap servers can also come from source descriptor.